package/MDAnalysis/core/groups.py

```diff
@@ -745,21 +767,73 @@ def isunique(self):
            >>> ag2.isunique
            True
 
+        See Also
+        --------
+
+        asunique
 
         .. versionadded:: 0.19.0
         """
         if len(self) <= 1:
             return True
-        # Fast check for uniqueness
-        # 1. get sorted array of component indices:
-        s_ix = np.sort(self._ix)
-        # 2. If the group's components are unique, no pair of adjacent values in
-        #    the sorted indices array are equal. We therefore compute a boolean
-        #    mask indicating equality of adjacent sorted indices:
-        mask = s_ix[1:] == s_ix[:-1]
-        # 3. The group is unique if all elements in the mask are False. We could
-        #    return ``not np.any(mask)`` here but using the following is faster:
-        return not np.count_nonzero(mask)
+        return unique_int_1d(self._ix).shape[0] == self._ix.shape[0]
+    
+    def _asunique(self, group, sorted=False, set_mask=False):
+        try:
+            name = 'sorted_unique' if sorted else 'unsorted_unique'
+            return self._cache[name]
+        except KeyError:
+            pass
+
+        if self.isunique:
+            if not sorted:
+                self._cache['unsorted_unique'] = self
+                return self
+            if self.issorted:
+                self._cache['unsorted_unique'] = self
+                self._cache['sorted_unique'] = self
+                return self
+
+        if sorted:
+            if 'unsorted_unique' in self._cache and self.unsorted_unique.issorted:
+                self._cache['sorted_unique'] = self.unsorted_unique
+                return self.unsorted_unique
```

Comment:
With all this complicated logic of what's cached, sorted, unique etc it would be good to have this `if` clause tested, too.